packages/core/src/application.ts

```diff
@@ -130,6 +131,9 @@ export class Application extends Context {
    * @memberof Application
    */
   public async start(): Promise<void> {
+    await this._forEachComponent(c => {
```

Comment:
This opens an interesting question: should we invoke "start" observers before the servers are started or after?

 - When the observers are invoked before, they have a chance to modify server configuration.
 - When the observers are invoked after, they are able to access the information created during startup, e.g. the port number assigned by the operation system to build the URL of a server listening on `port: 0`.

I think the second use case is more important because I don't see how it could be implemented otherwise? The first use case (modifying config before start) seems redundant to me, because application setup can already be customized during `app.boot()`.

Reply:
Please make sure there is a test capturing this nuance!

Reply:
It's always tricky to make sure artifacts are started/stopped in the right order. For example, we usually want servers to be started last and stopped first so that requests can only be handled when the system is ready.

A way to describe the dependency is needed to allow life cycle actions to be triggered by order. For example, each action can declare its dependencies or desired phase.